File: benchmark/autolifter/single-pass/mis.f
Config ExtraGrammar = "AutoLifter";

Inductive List = cons {Int, List} | nil Unit;

single_pass = \v: List -> Int.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

/*User provided programs*/

max = \a: Int. \b: Int. if < a b then b else a;

mis = fix (
  \f: Int -> Int -> List -> Int. \p: Int. \np: Int. \xs: List.
  match xs with
    nil _ -> max p np
  | cons {h, t} ->
    f (+ h np) (max p np) t
  end
) 0 0;

main = single_pass mis;

File: ml-benchmark/autolifter/single-pass/mis.ml
type l = Nil of unit | Cons of int * l

let single_pass v =
  let rec run xs =
    match xs with
    | Nil _ -> xs
    | Cons (h, t) -> Cons (h, run t)
  in
  fun xs -> v (run xs)

let max a b = if a < b then b else a

let rec mis_f p np xs =
  match xs with
  | Nil _ -> max p np
  | Cons (h, t) -> mis_f (h + np) (max p np) t

let mis xs = mis_f 0 0 xs

let program = single_pass mis